project/ian-chiu/README.md

```diff
@@ -0,0 +1,185 @@
+
+### Cryptography
+
+Celluar automata are proposed for use in public key cryptography. They can be utilized to construct pseudorandom number generators, and to design error correction codes.
+
+### Generative Art and Music
+
+Cellular automata have been used in generative music and evolutionary music composition and procedural terrain generation in video games.
+
+*Sources*:
+* Tech Target: https://www.techtarget.com/searchenterprisedesktop/definition/cellular-automaton
+* Wikipedia:
+https://en.wikipedia.org/wiki/Cellular_automaton#Applications
+
+## System Architecture
+
+### Features
+
+The software should have following features:
+
+1. Simulation operations: randomize, run, pause, clear
+2. Simulation settings: randomize density and range, grid size
+3. Some cell rules can be used by default
+4. Users can add their own cells rules
+5. Basic camera control: Orbiting and Zooming
+
+### Input / Output
+
+* Input: GUI, keyboard, mouse
+* Ouput: application window
+
+### Third Party Library
+
+- raylib-py or pygame: A simple video games programming library for dealing with windows, keyboard/mouse input, graphics and GUI.
+
+## API Description
+
+### Python
+
+* Rule: 
+    * apply: apply the cell rule
+    * count_neighbors: get number of alive neighbors according to the rule
+* CameraController:
+    * update: update camera view and projection matrix according to the control input
+* Application:
+    * run: control the render loop and the all update for the next frame
+    * randomize: randomize all cells in a grid
+    * clear: empty all cells in a grid
+
+### C++
+
+* calculateNextState: calculate all cells' states of the next frame. Should use multithreading and SIMD or any parallel computing techniques to increase computation speed.
+* getGreedyMeshes: get the merged meshes after executing the greedy meshing algorithm
+
+## Engineering Infrastructure
+
+|Category|Tools|
+|--|--|
+|Build System|CMake|
+|Testing Framework|pytest|
+|Version Control|Git|
+
+## Schedule
+
+### Planning phase (6 weeks from 9/19 to 10/31): 
+
+- Learn some parallel computing techniques, like multi-threading and SIMD operations. 
+- Get familiar with the greedy meshing algorithm and how to use the third party library.
+- Create the application window and get the render loop running. 
+- Render a cube on the window.
+
+### Week 1 (10/31): 
+
+- Make sure the keyboard and mouse input can work.
+- Finish camera controller.
+
+### Week 2 (11/7):
+
+- Setup the grid.
+- Implement cell rules in python.
+- Implement calculateNextCells in C++ without optimization.
+- Make sure cells in the grid behave correctly in every iteration according to the applying rule.
+- Write a python test for a simple cell rule.
+
+
+### Week 3 (11/14):
+
+- Create the GUI.
+- Implement all the simulation operations and settings that needs GUI.
+
+### Week 4 (11/21):
+
+- Optimize computing speed of calculateNextCells by using the parallel computing techniques, like SIMD or multi-threading. 
```

Comment:
I would be satisfied even if you simply optimize the code without parallelism.  Parallelism is good to have.

Reply:
Ok, good to know.